react-components/src/query/useAssetsAndTimeseriesLinkageDataQuery.ts

```diff
@@ -0,0 +1,219 @@
+ */
+import { useQuery, type UseQueryResult } from '@tanstack/react-query';
+
+import {
+  type Asset,
+  type ExternalId,
+  type IdEither,
+  type InternalId,
+  type Timeseries,
+  type CogniteExternalId,
+  type RelationshipResourceType,
+  type CogniteClient
+} from '@cognite/sdk';
+
+import { getRelationships } from '../hooks/network/getRelationships';
+import { useSDK } from '../components/RevealCanvas/SDKProvider';
+import {
+  type AssetIdsAndTimeseries,
+  type AssetAndTimeseriesIds,
+  type RelationshipsFilterInternal,
+  type AssetIdsAndTimeseriesData
+} from '../utilities/types';
+import { createLabelFilter } from '../utilities/createLabelFilters';
+import { queryKeys } from '../utilities/queryKeys';
+import { getTimeseriesByIds } from '../hooks/network/getTimeseriesByIds';
+import { isDefined } from '../utilities/isDefined';
+import { getAssetsByIds } from '../hooks/network/getAssetsByIds';
+import { getTimeseriesLatestDatapoints } from '../hooks/network/getTimeseriesLatestDatapoints';
+
+type Props = {
+  timeseriesExternalIdsFromRule: CogniteExternalId[];
+  contextualizedAssetNodes: Asset[];
+  filter?: RelationshipsFilterInternal;
+};
+
+export function useAssetsAndTimeseriesLinkageDataQuery({
+  timeseriesExternalIdsFromRule,
+  contextualizedAssetNodes,
+  filter
+}: Props): UseQueryResult<AssetIdsAndTimeseriesData, unknown> {
+  const sdk = useSDK();
+
+  const relationshipResourceTypes: RelationshipResourceType[] = ['asset'];
+
+  return useQuery(
+    [
+      queryKeys.timeseriesLinkedToAssets(),
+      timeseriesExternalIdsFromRule,
+      relationshipResourceTypes,
+      filter
+    ],
+    async () => {
+      const assetAndTimeseriesIdsFromRelationship = await getLinkFromRelationships(
+        sdk,
+        timeseriesExternalIdsFromRule,
+        relationshipResourceTypes,
+        filter
+      );
+
+      const externalIds: ExternalId[] = timeseriesExternalIdsFromRule.map((externalId) => {
+        return {
+          externalId
+        };
+      });
+
+      const timeseries = await getTimeseriesByIds(sdk, externalIds);
+
+      const assetIdsFound: IdEither[] =
+        timeseries
+          ?.map((timeseries) => {
+            return getAssetIdsFromTimeseries(
+              contextualizedAssetNodes,
+              timeseries,
+              assetAndTimeseriesIdsFromRelationship
+            );
+          })
+          .flat()
+          .filter(isDefined) ?? [];
+
+      const assetFromTimeseries = await getAssetsByIds(sdk, assetIdsFound);
+      const assetIdsWithTimeseries =
+        timeseries
+          ?.map((timeseries): AssetIdsAndTimeseries[] | undefined => {
+            return generateAssetAndTimeseries(
+              assetFromTimeseries,
+              timeseries,
+              assetAndTimeseriesIdsFromRelationship
+            );
+          })
+          .flat()
+          .filter(isDefined) ?? [];
+
+      const timeseriesDatapoints = await getTimeseriesLatestDatapoints(sdk, externalIds);
```

Comment:
I see that we do several `await`s serially in this function. Can/should we await them in parallel?

Reply:
Awesome! Now we're talking ahaha nice catch! Like it!
Improved here.
https://github.com/cognitedata/reveal/pull/4414/commits/0788178dd8d724640516e763f63ca14c51e0cb59

Reply:
Nice!